### ISSUE TITLE
React-Native appends "charset=utf-8" to HTTP call Content-Type

### ISSUE THREAD

**Issue by laurent22** [facebook/react-native#14445]
### Description

I'm trying to use the Dropbox API with React-Native but when I try to call [filesUpload](https://dropbox.github.io/dropbox-sdk-js/Dropbox.html#filesUpload__anchor) I'm getting the following error:

	Error in call to API function "files/upload": Bad HTTP "Content-Type" header: "application/octet-stream; charset=utf-8".  Expecting one of "application/octet-stream", "text/plain; charset=dropbox-cors-hack".

The Dropbox SDK makes use of the [superagent](https://github.com/visionmedia/superagent) library and it sets up the call like this:

    apiRequest = request.post(getBaseURL(host) + path)
      .type('application/octet-stream')
      .set('Authorization', 'Bearer ' + accessToken)
      .set('Dropbox-API-Arg', httpHeaderSafeJson(args));

So the content-type is set to just "application/octet-stream" but when the request is eventually executed it's changed to "application/octet-stream; charset=utf-8". I'm assuming the SDK works fine in Node or in a browser, and that it's something that React-Native is adding. Is it a known issue? Is there any way to go around it?

### Solution

Make sure that HTTP calls coming from React-Native do not append `; charset=utf-8` to the Content-Type (or make it optional).

### Additional Information

* React Native version: 0.44
* Platform: Android (running on emulator under Windows)
* Development Operating System: Windows 10
* Dev tools: Android Studio

**Comment by hramos** [comment]
As this involves a third party library, can you try and reproduce this using `fetch()`? If RN is modifying the request in the way you describe, then the behavior can likely be observed using `fetch()`.

**Comment by laurent22** [comment]
After some digging, it seems to be related to the okhttp library. I've put my findings in this related issue:

https://github.com/facebook/react-native/issues/8237#issuecomment-307578320

I've also added a comment to the okhttp issue but it doesn't look like they are going to fix it:

https://github.com/square/okhttp/issues/3081#issuecomment-307581489

**Comment by ericwooley** [comment]
I am having the same issue using super agent directly, in emulators and on device. It doesn't occur in ios.

If i find a solution I will post it here

**Comment by dauane** [comment]
I have the same issue. Tried using different third party libraries and even RN fetch. It all failed.

**Comment by dauane** [comment]
Fixing the server was the simplest solution for me. Thanks @ericwooley

**Comment by hramos** [comment]
Hi there! This issue is being closed because it has been inactive for a while. Maybe the issue has been fixed in a recent release, or perhaps it is not affecting a lot of people. Either way, we're automatically closing issues after a period of inactivity. Please do not take it personally!

If you think this issue should definitely remain open, please let us know. The following information is helpful when it comes to determining if the issue should be re-opened:

- Does the issue still reproduce on the latest release candidate? Post a comment with the version you tested.
- If so, is there any information missing from the bug report? Post a comment with all the information required by the [issue template](https://github.com/facebook/react-native/blob/master/.github/ISSUE_TEMPLATE.md).
- Is there a pull request that addresses this issue? Post a comment with the PR number so we can follow up.

If you would like to work on a patch to fix the issue, *contributions are very welcome*! Read through the [contribution guide](http://facebook.github.io/react-native/docs/contributing.html), and feel free to hop into [#react-native](https://discordapp.com/invite/0ZcbPKXt5bZjGY5n) if you need help planning your contribution.

**Comment by laurent22** [comment]
@hramos , this is still a **major** issue for me and quite a few others. The Dropbox API still cannot be used from React Native, which prevents me from adding support for it in my app. It would be good if the issue could be re-opened. It's still not working in the latest release candidate and as far as I know there's currently no pull request addressing the issue.

**Comment by Kottidev** [comment]
@hramos , Our server does not prepare api for “utf-8”, and android avd sent requests with “charset=utf-8". So it cause 500 status error.

the issue should be re-opened

**Comment by tastafur** [comment]
In our react-native application we have the same problem

**Comment by ferroblesh** [comment]
same thing here. It's incredible that react native claims to be same core and uses different kind of http request for each platform. It should be same behavior on both sides

**Comment by lil5** [comment]
Here's a workaround https://github.com/facebook/react-native/issues/13487#issuecomment-319984503

**Comment by Schonhoffer** [comment]
For anyone else stuck, here's what I did to get around it.

`npm install --save buffer`

`import { Buffer } from 'buffer'`

```javascript
let content = "foo";
let resp = await fetch(
            'https://content.dropboxapi.com/2/files/upload', {
            method: 'POST',
            body:  Buffer.from(content, 'utf8'),
            headers: new Headers({
              'Authorization': 'Bearer ' + this._accessToken,
              'Dropbox-API-Arg': JSON.stringify(args),
              'Content-Length' : content.length,
              'Content-Type': 'application/octet-stream'
            })}
        );
```

**Comment by Dubidubiduu** [comment]
Please can someone solves this bug? Waiting for nextcloud support which isn't developed before closing this bug: http://joplin.cozic.net/

**Comment by vovkasm** [comment]
@Schonhoffer excellent!
Buffer will use ArrayBuffer, then
in [convertRequestBody.js](https://github.com/facebook/react-native/blob/0.52-stable/Libraries/Network/convertRequestBody.js#L35) it will be encoded as base64 to marshal JS->Java, finally
in [NetworkingModule.java](https://github.com/facebook/react-native/blob/9a8784c241464c581c3ca8584511244666f9efb5/ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java#L270) it will properly reach OkHttp as `ByteString`!

@Dubidubiduu try solution from @Schonhoffer it should work.

**Comment by aeternas** [comment]
Is the problem still reproduced after this fix? https://github.com/facebook/react-native/pull/23580#issuecomment-466291015

**Comment by dulmandakh** [comment]
The issue is fixed in 0.60.

**Comment by jennifer-elyse** [comment]
> For anyone else stuck, here's what I did to get around it.
> 
> `npm install --save buffer`
> 
> `import { Buffer } from 'buffer'`
> 
> ```js
> let content = "foo";
> let resp = await fetch(
>             'https://content.dropboxapi.com/2/files/upload', {
>             method: 'POST',
>             body:  Buffer.from(content, 'utf8'),
>             headers: new Headers({
>               'Authorization': 'Bearer ' + this._accessToken,
>               'Dropbox-API-Arg': JSON.stringify(args),
>               'Content-Length' : content.length,
>               'Content-Type': 'application/octet-stream'
>             })}
>         );
> ```

You can also add 
; charset=UTF-8 
to the Content-Type
